### PR TITLE
Remove absorption resonances in AlignAndFocusPowder

### PR DESCRIPTION
**Pull request by stonecoldhughes**
**Description of work.**

Add support for filtering out ranges of wavelengths - for filtering out resonance absorption peaks.

PR ---> ornl-next: https://github.com/mantidproject/mantid/pull/32063

**To test:**

```sh
./bin/WorkflowAlgorithmsTest AlignAndFocusPowderTest testEventWksp_filterResonance
```

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/WorkflowAlgorithms/src/AlignAndFocusPowder.cpp [hunk 1]
```diff
@@ -161,6 +166,19 @@ void AlignAndFocusPowder::init() {
...
   mustBePositive->setLower(0.0);
+
+  std::vector<std::string> allowedResonanceUnits({"Energy", "Wavelength"});
+  declareProperty(PropertyNames::RESONANCE_UNITS, allowedResonanceUnits.back(),
+                  std::make_shared<StringListValidator>(allowedResonanceUnits),
+                  "Units for resonances to be filtered in. "
+                  "The data will be converted to these units temporarily to filter.");
+  declareProperty(std::make_unique<ArrayProperty<double>>(PropertyNames::RESONANCE_LOWER_LIMITS),
+                  "Minimum values to filter absorption resonance. This must have same number of values as "
+                  "ResonanceFilterUpperLimits. Default behavior is to not filter.");
```
Comment (wdzhou) [inline Framework/WorkflowAlgorithms/src/AlignAndFocusPowder.cpp:177]: Could you be more specific about what is 'default behavior'?  Users of this method may get a little bit confused.

File: Framework/WorkflowAlgorithms/src/AlignAndFocusPowder.cpp [hunk 2]
```diff
@@ -239,6 +257,31 @@ std::map<std::string, std::string> AlignAndFocusPowder::validateInputs() {
                                         "Please plot the pCharge-time to identify suitable range for "
                                         "re-time-slicing";
 
+  m_resonanceLower = getProperty(PropertyNames::RESONANCE_LOWER_LIMITS);
+  m_resonanceUpper = getProperty(PropertyNames::RESONANCE_UPPER_LIMITS);
+  if (m_resonanceLower.size() > 0 || m_resonanceUpper.size() > 0) {
```
Comment (wdzhou) [inline Framework/WorkflowAlgorithms/src/AlignAndFocusPowder.cpp:262]: I don't think you need this if-statement.  If either of them is zero or both of them are zero, the codes in the condition block shall handle it well.

File: Framework/WorkflowAlgorithms/src/AlignAndFocusPowder.cpp [hunk 3]
```diff
@@ -434,15 +477,15 @@ void AlignAndFocusPowder::exec() {
   }
 
   // set up a progress bar with the "correct" number of steps
-  m_progress = std::make_unique<Progress>(this, 0., 1., 21);
+  m_progress = std::make_unique<Progress>(this, 0., 1., 22);
```
Comment (wdzhou) [inline Framework/WorkflowAlgorithms/src/AlignAndFocusPowder.cpp:480]: why 22?  any reason for this magic number?

Comment (peterfpeterson) [reply Framework/WorkflowAlgorithms/src/AlignAndFocusPowder.cpp:480]: There were 21 steps before this functionality, adding this makes one more (22=21+1)

**Review by wdzhou — APPROVED** [review]
It looks good to me. :shipit:

**Review by jmborr — DISMISSED** [review]
It's customary to provide a Workbench (manual) test for the reviewer to run, if only to make sure the reviewer built the feature branch 😆

Is a manual test unnecessary here?

**Comment by peterfpeterson** [comment]
Manual test is not entirely needed. The new unit test checks the behavior was done correctly.

**Pull request merged**